contents/security/sqlInjection.md

```diff
@@ -0,0 +1,217 @@
+```
+
+In SQL, `AND` operations are checked before `OR` operations. This query will check 
+the database for entries where:
+
+```
+Username = foo AND Password = bar
+OR
+‘’=’’
+```
+
+This query will always return true, as ‘’=’’ is always true. 
+As such, the query can be simplified to this:
+
+```sql
+SELECT * FROM Users
+```
+
+This will return all the rows from the Users table in the database, regardless of username 
+or password entered. If used in user authentication, the user will likely be able to gain
+access to the system.
+
+
+## Preventing SQL Injection ##
+
+There are a couple of ways to protect your website against SQL injection attacks. The two most
+common ways are:
+
+### Prepared Statements using Parameterized Queries ###
+
+By defining all the SQL code first and passing in the parameters afterwards, you can make 
+the database distinguish the difference between code and data. It would treat the values 
+entered by the user as a parameter, and would not allow it to alter the query being 
+executed. The way to achieve this varies by language, but it is easy to implement and effective.
+
+For example, instead of writing this in Java:
+
+```java
+String query = "SELECT * FROM Users WHERE Username = " + input_username +"’ 
+                AND Password = ‘" + input_password + "’";
+Statement statement = connection.createStatement();
+ResultSet results = statement.executeQuery( query );
+```
+
+You can prepare the statement like this:
+
+```java
+ String query = "SELECT * FROM Users WHERE Username = ? 
+                AND Password = ? ";
+ PreparedStatement pstmt = connection.prepareStatement( query );
+ pstmt.setString( 1, input_username ); 
+ pstmt.setString( 2, input_password );
+ ResultSet results = pstmt.executeQuery( );
+```
+
+The "?" in the query string is a placeholder for a string value. In the example, the first "?" 
+is substituted with the value of `input_username` using the call `pstmt.setString(1, input_username)`. 
+If the earlier attack was attempted, The query being submitted to the database will look like this:
+
+```sql
+SELECT * FROM Users WHERE Username = 'foo' AND Password = 'bar' OR ‘’=’’
+```
+
+However, since it is treated as a value to be used, the database will not allow it to modify the 
+query, and it would not be able to affect the structure of the SQL statement. As such, 
+the query will check the database for entries where:
+
+```
+Username = foo
+AND 
+Password = bar OR ‘’=’’
+```
+
+As such, the query is safe from external modifications.
+
+
+### Whitelist Input Validation ###
+
+By applying a whitelist to the values a user is allowed to use, you can 
+remove undesired symbols in the query being passed to the database 
+(e.g. whitelisting only alphanumeric characters for a username).
+This ensures that attackers won't be allowed to enter special characters 
+which may have unwanted effects when executed. 
+(e.g. special characters in SQL such as `'`, `@`, `^` and `_`)
```

Comment:
Remove these extra line breaks. We do not require markdown text to wrap at a certain length.